### PR TITLE
fix(api): fix converter output against schema

### PR DESCRIPTION
**Pull request by Orta21**
Ref. metriport/metriport#799

Ticket: #799

### Dependencies

- Upstream: none
- Downstream: none

### Description

Patch issues with FHIR converter output against schema. 

### Testing

- Local
  - [x] Validate bundle outputs
- Staging
  - [ ] Validate bundle outputs
- Production
  - [ ] Validate bundle outputs

Check each PR.

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js [hunk 1]
```diff
@@ -1154,7 +1165,7 @@ module.exports.external = [
       if (referenceData == undefined) {
         return "";
       }
-      return JSON.stringify(referenceData).slice(1, -1);
+      return JSON.stringify(referenceData).slice(1, -1).replace(/\s+/g, "").trim();
```
Comment (Orta21) [inline packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:1168]: The regex is to remove double spaces (it breaks schema)

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js [hunk 2]
```diff
@@ -1336,7 +1347,19 @@ module.exports.external = [
...
       }
       const match = str.match(new RegExp(`^(${DECIMAL_REGEX_STR})$`));
-      return match ? match[0] : "";
+
+      if (match) {
+        const decimal = match[0];
+        const leadsWithDecimal = decimal.startsWith(".");
+
+        if (leadsWithDecimal) {
+          return `0${decimal}`;
```
Comment (Orta21) [inline packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:1356]: If we dont add a zero in front of a leading decimal it breaks the structure of the json

File: packages/fhir-converter/src/templates/cda/Header.hbs [hunk 3]
```diff
@@ -79,14 +79,16 @@
...
+
+    {{#each (toArray document.author) as |author|}}
+      {{#if author}}
+          {{#with (evaluate 'Utils/GenerateId.hbs' obj=author.assignedAuthor) as |practitionerId|}}
+            {{>Resources/Practitioner.hbs practitioner=author.assignedAuthor ID=practitionerId.Id}},
             {{>References/Composition/author.hbs ID=(generateUUID (toJsonString document)) REF=(concat 'Practitioner/' practitionerId.Id)}},
-        {{/with}}
-    {{/if}}
-        
+          {{/with}}
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Header.hbs:88]: this is probably why we have more practitioners - was this intentional to move to looping through all authors?

File: packages/fhir-converter/src/templates/cda/Sections/Payers.hbs [hunk 4]
```diff
@@ -40,7 +40,7 @@
                     Example: https://github.com/HL7/C-CDA-Examples/blob/fb1c33b3a880fc47407f25c5dafa0066996529f4/Documents/CCD/CCD%201/CCD.XML
                  --}}
                     {{#if (contains (toJsonString performerEntry.templateId) "2.16.840.1.113883.10.20.22.4.87")}}
-                        {{#if performerEntry.assignedEntity.representedOrganization.name._}}
+                        {{#if performerEntry.assignedEntity.representedOrganization}}
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Sections/Payers.hbs:43]: why was this changed? won't we just end up with a bunch of useless/nameless org objects now?

File: packages/fhir-converter/src/templates/cda/Resources/AllergyIntolerance.hbs [hunk 5]
```diff
@@ -48,25 +48,32 @@
...
+                          {{/if}}
+                      {{/if}}
+
+                  {{/each}}
+                {{else}}
+                  "manifestation":
+                  [
+                      {{>DataType/CodeableConcept.hbs text='noKnownManifestation'}},
+                  ],
+                {{/if}}
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Resources/AllergyIntolerance.hbs:76]: general feedback - let's make "unknown" things to spec, and standardized with our mapping so that our customers can have a consistent way of determining something is known

suggestion is to use `UNK` in https://terminology.hl7.org/1.0.0/ValueSet-v3-Unknown.html for all CodeableConcepts / Codings that we don't have enough info on to fill out

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js (hunk 1 above)
Comment (jonahkaye) [reply packages/fhir-converter/src/lib/handlebars-converter/handlebars-helpers.js:1168]: This removes all whitespaces from a string

`replace(/ {2,}/g, " ")` would replace doubles with singles

File: packages/fhir-converter/src/templates/cda/DataType/Coding.hbs [hunk 6]
```diff
@@ -25,7 +25,7 @@
   // -------------------------------------------------------------------------------------------------
 --}}
 {
-	"code":"{{code.code}}",
+	"code":"{{trim code.code}}",
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/DataType/Coding.hbs:28]: what is the issue this is fixing?

File: packages/fhir-converter/src/templates/cda/Header.hbs [hunk 7]
```diff
@@ -79,14 +79,16 @@
             {{/with}}           
         {{/if}}
     {{/if}}
-            
-    {{#if document.author.assignedAuthor}}
-        {{#with (evaluate 'Utils/GeneratePractitionerId.hbs' obj=document.author.assignedAuthor) as |practitionerId|}}
-            {{>Resources/Practitioner.hbs practitioner=document.author.assignedAuthor ID=practitionerId.Id}},
+
+    {{#each (toArray document.author) as |author|}}
+      {{#if author}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Header.hbs:84]: we removed the check here if assignedAuthor is defined. 
this will now create even if its undefined and result in us creating empty practitioners. Lets add it back to be if author.assignedAuthor

File: packages/fhir-converter/src/templates/cda/Resources/AllergyIntolerance.hbs [hunk 8]
```diff
@@ -48,25 +48,32 @@
...
-                    {{/if}}
-                    
-                {{/each}}
+                {{#if allergyEntry.entryRelationship }}
+                  {{#each (toArray allergyEntry.entryRelationship)}}
+
+                      {{#if (contains (toJsonString this.observation.templateId) '"2.16.840.1.113883.10.20.22.4.9"')}}
+                          "manifestation":
+                          [
+                              {{>DataType/CodeableConcept.hbs code=this.observation.code}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/AllergyIntolerance.hbs:57]: value -> code
why?

File: packages/fhir-converter/src/templates/cda/Resources/Consent.hbs [hunk 9]
```diff
@@ -44,7 +44,12 @@
               }
             ]
         }
-        "category":[{{>DataType/CodeableConcept.hbs code=categoryCode}}],    
+
+        {{#if categoryCode.code}}
+          "category":[{{>DataType/CodeableConcept.hbs code=categoryCode}}],
+        {{else}}
+          "category":[{{>DataType/CodeableConcept.hbs text='n/a'}}],
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Consent.hbs:51]: ditto on dimas comment about unknown things being to spec

File: packages/fhir-converter/src/templates/cda/Resources/DiagnosticReport.hbs [hunk 10]
```diff
@@ -36,7 +36,11 @@
             {{/each}}
         ],
         "status":{{>ValueSet/DiagnosticReportStatus.hbs code=diagReport.statusCode.code}},
-        "code":{{>DataType/CodeableConcept.hbs code=diagReport.code}},
+        {{#if diagReport.code.code}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/DiagnosticReport.hbs:39]: bug? code.code
Looks wrong since within a code, the code might be in the translation etc.

File: packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs [hunk 11]
```diff
@@ -46,9 +46,19 @@
...
+
+        {{#if encounter.code.code}}
+            "class":{{>DataType/Coding.hbs code=encounter.code}},
+        {{else}}
+          {{#if encounter.code.translation}}
+            "class":{{>DataType/Coding.hbs code=encounter.code.translation }},
+          {{else}}
+            "class": {
+              "display": "n/a"
+            }
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs:58]: why dont you just update coding.ts to display the unk value if there was no defined code in the code passed to it and then we can leave the template as it was before? here and elsewhere

File: packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs [hunk 12]
```diff
@@ -36,7 +36,12 @@
             {{/each}}
         ],
         "status":"completed"
-        "relationship":{{>DataType/CodeableConcept.hbs code=familyEntry.subject.relatedSubject.code}},
+        
+        {{#if familyEntry.subject.relatedSubject.code}}
+          "relationship":{{>DataType/CodeableConcept.hbs code=familyEntry.subject.relatedSubject.code}},
+        {{else}}
+          "relationship":{{>DataType/CodeableConcept.hbs text='n/a'}},
+        {{/if}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs:44]: ditto on updating CodeableConcept itself and not changing template here

File: packages/fhir-converter/src/templates/cda/Resources/Observation.hbs [hunk 13]
```diff
@@ -60,8 +60,17 @@
             {{/if}}
         
         "status":{{>ValueSet/ObservationStatus.hbs code=observationEntry.statusCode.code}},
-        "code": {{>DataType/CodeableConcept.hbs code=observationEntry.code text=observationEntry.text._}},
-        
+
+        {{#if observationEntry.code.code}}
+          "code": {{>DataType/CodeableConcept.hbs code=observationEntry.code text=observationEntry.text._}},
+        {{else}}
+          {{#if observationEntry.value.length}}
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Observation.hbs:67]: i am not familiar with the length property of value. Is it actually in one of these:
https://build.fhir.org/ig/HL7/CDA-core-2.0/StructureDefinition-Observation-definitions.html#diff_Observation.value

File: packages/fhir-converter/src/templates/cda/Resources/Observation.hbs [hunk 14]
```diff
@@ -60,8 +60,17 @@
...
+            "code": {{>DataType/CodeableConcept.hbs code=observationEntry.value.[0] text=observationEntry.text._}},
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Observation.hbs:68]: 0 indexing but what if its not an array or empty?

File: packages/fhir-converter/src/templates/cda/Header.hbs (hunk 3 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Header.hbs:88]: A Composition cannot be created without an author (required by schema). When i logged the document.author for xmls that had issue its because it was an array. So i looped through it to get all the assignedAuthors.

File: packages/fhir-converter/src/templates/cda/Sections/Payers.hbs (hunk 4 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Sections/Payers.hbs:43]: The problem is Coverage needs a payor (OrgRef). So i created the nameless Org so it has something to reference. What I could do instead is not create it and if the Org doesnt exist make the payor like this

`        
"payor": [
    {
    "reference": "UNK"
    }
],`

File: packages/fhir-converter/src/templates/cda/DataType/Quantity.hbs
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/DataType/Quantity.hbs:1]: Because it breaks the FHIR validation. It needs to be of type  `"$ref": "#/definitions/decimal"`

File: packages/fhir-converter/src/templates/cda/DataType/Coding.hbs (hunk 6 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/DataType/Coding.hbs:28]: code cant have trailing spaces

`"description": "A string which has at least one character and no leading or trailing whitespace and where there is no whitespace other than single spaces in the contents"`

File: packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs (hunk 11 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs:58]: I thought about it but we only want it adding unknown values for fields that are mandatory. If we add this logic to `CodeableConcept` we will render a lot more unknowns than we need too

File: packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs (hunk 12 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Resources/FamilyMemberHistory.hbs:44]: https://github.com/metriport/metriport/pull/2357#discussion_r1662607267

File: packages/fhir-converter/src/templates/cda/Resources/Observation.hbs (hunk 13 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Resources/Observation.hbs:67]: I applied the same logic as DiagnosticReport so removed the value logic

File: packages/fhir-converter/src/templates/cda/Resources/Observation.hbs (hunk 14 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Resources/Observation.hbs:68]: ditto removed this

File: packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs (hunk 11 above)
Comment (jonahkaye) [reply packages/fhir-converter/src/templates/cda/Resources/Encounter.hbs:58]: You could add a new CodeableConcept primitive for when you want it to use unknowns if not defined. Or pass in a boolean flag etc

File: packages/fhir-converter/src/templates/cda/Resources/Consent.hbs [hunk 15]
```diff
@@ -48,7 +48,7 @@
         {{#if categoryCode.code}}
           "category":[{{>DataType/CodeableConcept.hbs code=categoryCode}}],
         {{else}}
-          "category":[{{>DataType/CodeableConcept.hbs text='n/a'}}],
+          "category":[{{>DataType/CodeableConcept.hbs code='UKN' text='unknown'}}],
```
Comment (jonahkaye) [inline packages/fhir-converter/src/templates/cda/Resources/Consent.hbs:51]: nit: UKN -> UNK

File: packages/fhir-converter/src/templates/cda/DataType/Quantity.hbs
[the remark this answers is not in the document]
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/DataType/Quantity.hbs:1]: All Quantiy Values are decimals 

`
        "value": {
          "description": "The value of the measured amount. The value includes an implicit precision in the presentation of the value.",
          "$ref": "#/definitions/decimal"
        },
  `
  
  Updated `extractDecimal` to return a float

**Review by jonahkaye — APPROVED** [review]

File: packages/fhir-converter/src/templates/cda/DataType/Coding.hbs [hunk 16]
```diff
@@ -25,7 +25,7 @@
   // -------------------------------------------------------------------------------------------------
 --}}
 {
-	"code":"{{code.code}}",
+	"code": {{#if code.code}} "{{trim code.code}}" {{else if canBeUnknown}} "UNK" {{/if}},
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/DataType/Coding.hbs:28]: based on my previous comment - can't we fill out the `code`, `system`, and `display` in the `UNK` case?

File: packages/fhir-converter/src/templates/cda/Sections/Payers.hbs [hunk 17]
```diff
@@ -40,12 +40,14 @@
...
                     {{#if (contains (toJsonString performerEntry.templateId) "2.16.840.1.113883.10.20.22.4.87")}}
-                        {{#if performerEntry.assignedEntity.representedOrganization.name._}}
+                          {{#if performerEntry.assignedEntity.representedOrganization.name._}}
                             {{#with (evaluate 'Utils/GenerateOrganizationId.hbs' obj=performerEntry.assignedEntity.representedOrganization) as |orgId|}}
                                 {{>Resources/Organization.hbs org=performerEntry.assignedEntity.representedOrganization ID=orgId.Id}},
                                 {{>References/Coverage/payor.hbs ID=(generateUUID (toJsonString payerEntry.act)) REF=(concat 'Organization/' orgId.Id)}},
                                 {{>References/Coverage/beneficiary.hbs ID=(generateUUID (toJsonString payerEntry.act)) REF=(concat 'Patient/' @metriportPatientId)}},
                             {{/with}}
+                        {{else}}
+                          {{>References/Coverage/payor.hbs ID=(generateUUID (toJsonString payerEntry.act)) REF='UNK' }},
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/Sections/Payers.hbs:50]: won't this result in references like `Payor/UNK` ?

**Review by Goncharo — CHANGES_REQUESTED** [review]

File: packages/fhir-converter/src/templates/cda/Sections/Payers.hbs (hunk 17 above)
Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Sections/Payers.hbs:50]: Yea thats what i mentioned here https://github.com/metriport/metriport/pull/2357#discussion_r1662496684

Comment (Orta21) [reply packages/fhir-converter/src/templates/cda/Sections/Payers.hbs]: Addressed so instead if its unknown it wont have a reference but rather a display with uknown

File: packages/fhir-converter/src/templates/cda/DataType/Coding.hbs [hunk 18]
```diff
@@ -25,10 +25,14 @@
...
 	"version":"{{code.codeSystemVersion}}",
-	{{#with (evaluate 'ValueSet/SystemReference.hbs' code=code.codeSystem) as |system|}}
-		"system": "{{system.oid}}",
-	{{/with}}
+  {{#if code.codeSystem}}
+    {{#with (evaluate 'ValueSet/SystemReference.hbs' code=code.codeSystem) as |system|}}
+      "system": "{{system.oid}}",
+    {{/with}}
+  {{else if canBeUnknown}}
+    "system": "2.16.840.1.113883.1.11.10612",
```
Comment (Goncharo) [inline packages/fhir-converter/src/templates/cda/DataType/Coding.hbs:36]: why don't we just put the URL here so it's obvious/navigatable at a glance?

**Review by Goncharo — CHANGES_REQUESTED** [review]

**Review by Goncharo — APPROVED** [review]
:shipit:
